report/marshal.go

```diff
@@ -0,0 +1,53 @@
+package report
+
+import (
+	"compress/gzip"
+	"encoding/gob"
+	"io"
+
+	"github.com/ugorji/go/codec"
+)
+
+// WriteBinary writes a Report as a gzipped msgpack.
+func (rep Report) WriteBinary(w io.Writer) error {
+	gzwriter, err := gzip.NewWriterLevel(w, gzip.BestCompression)
+	if err != nil {
+		return err
+	}
+	if err = codec.NewEncoder(gzwriter, &codec.MsgpackHandle{}).Encode(&rep); err != nil {
+		return err
+	}
+	gzwriter.Close() // otherwise the content won't get flushed to the output stream
+	return nil
+}
+
+// ReadBinary reads into a Report from a gzipped msgpack.
+//
+// Will decompress the binary if gzipped is true, and will use the given
+// codecHandle to decode it. If codecHandle is nil, will decode as a gob.
+func (rep *Report) ReadBinary(r io.Reader, gzipped bool, codecHandle codec.Handle) error {
+	var err error
+	if gzipped {
+		r, err = gzip.NewReader(r)
+		if err != nil {
+			return err
+		}
+	}
+	decoder := gob.NewDecoder(r).Decode
```

Comment:
I know it was like this before, but... we are wasting memory allocation here if codecHandle != nil. I suggest paying the price of an extra 'var' and 'else' to avoid that.